[analyze] Add Analyzer interface for Gitlab

abmussani opened the pull request:
### Description:
This PR implements the analyzer interface for Gitlab. 

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

abasit-folio3 committed:
implement analyzer interface for gitlab

abasit-folio3 committed:
generated permissions, added unit test for gitlab analyzer

abasit-folio3 committed:
revert deletion of scopes.go

abmussani committed:
appending domain in resource names

abmussani committed:
[chore]
moved expected output of test in json file to neat the code.

mcastorina reviewed (commented):
The code looks good but the test isn't passing. It looks like the gitlab token is expired:

```
» curl 'https://gitlab.com/api/v4/personal_access_tokens/self' -H "Private-Token: $GITLAB_TOKEN"
{
  "error": "invalid_token",
  "error_description": "Token is expired. You can either do re-authorization or token refresh."
}

```

abmussani committed:
updated the test for gitlab analyzer
to make more unique FullyQualifiedName, Ids are added for resources.

mcastorina commented on pkg/analyzer/analyzers/gitlab/gitlab.go:
```diff
@@ -26,18 +29,80 @@ type Analyzer struct {
+		FullyQualifiedName: fmt.Sprintf("github.com/token/%d", info.AccessToken.ID),
+		Type:               "access_token",
+		Metadata: map[string]any{
+			"last_used_at": info.AccessToken.LastUsedAt,
```
Since we don't update resources, this value can easily become stale. I think it's best to remove it.

abmussani replied:
removed.

mcastorina commented on pkg/analyzer/analyzers/gitlab/gitlab.go:
```diff
@@ -26,18 +29,80 @@ type Analyzer struct {
+	for _, project := range info.Projects {
+		projectResource := analyzers.Resource{
+			Name:               project.NameWithNamespace,
+			FullyQualifiedName: fmt.Sprintf("github.com/project/%d", project.ID),
```
Same here `github.com` -> `gitlab.com`

abmussani replied:
updated 🫡

abmussani committed:
remove unnecessary metadata field and fix github -> gitlab

abmussani committed:
Merge branch 'main' into impl-data-model-gitlab

abmussani commented on pkg/analyzer/analyzers/gitlab/gitlab.go:
```diff
@@ -26,18 +29,80 @@ type Analyzer struct {
+		Bindings: []analyzers.Binding{},
+	}
+
+	// Add token and it's permissions to bindings
```
Yup, The permission determine what action can be perform using that token.

mcastorina replied:
Permissions operate on a resource though, so in this case the permissions would describe operations that could be performed on the token itself, which seems wrong to me..

abasit-folio3 replied:
@mcastorina I had gone through the login. PAT token does belongs to user and we do get `userID` in response. So, resource can be a User with those permissions. What do you think ?

abmussani replied:
@mcastorina I had gone through the login. PAT token does belongs to user and we do get userID in response. So, resource can be a User with those permissions. What do you think ?

abmussani committed:
extract user id from access token json, make user as resource

abmussani committed:
link analyzer with gitlab v2 detector

mcastorina reviewed (approved):
lgtm!

Some of the tests are failing though - I think it's because we removed `analyzerpb`? You might have to update your branch to fix those errors

abmussani committed:
Merge branch 'main' into impl-data-model-gitlab

* main: (76 commits)
  update aws descriptions (#3529)
  enforce timeout on circleci test (#3528)
  rm snifftest (#3527)
  Redact more source credentials (#3526)
  Create global log redaction capability (#3522)
  Adding basic "what is trufflehog" to the readme (#3514)
  Handle custom detector response and include in extra data (#3411)
  fix: fixed validation logic for `calendarific` (#3480)
  fix(deps): update github.com/tailscale/depaware digest to 3d7f3b3 (#3518)
  Move DecoderType into ResultWithMetadata #3502
  Addeded 403 account block status code handling for gitlab (#3471)
  updated gcpapplicationdefaultcredentials detector results with RawV2 (#3499)
  fix(deps): update module github.com/brianvoe/gofakeit/v7 to v7.1.1 (#3512)
  fix(deps): update module github.com/schollz/progressbar/v3 to v3.17.0 (#3510)
  fix(deps): update module cloud.google.com/go/secretmanager to v1.14.2 (#3498)
  Adds a logging section in the contributing guidelines (#3509)
  fix: fixed verifcation pattern logic for `bulksms` (#3478)
  Extend `algoliaadminkey` with additional checks (#3459)
  fix(deps): update module google.golang.org/api to v0.203.0 (#3497)
  fix: added correct api endpoint for verification & logic for Aeroworkflow (#3435)
  ...

abmussani committed:
fixed code breaking changes due to analyzer protobuf removal.

abmussani committed:
Merge branch 'main' into impl-data-model-gitlab

abmussani commented:
> lgtm!
> 
> Some of the tests are failing though - I think it's because we removed `analyzerpb`? You might have to update your branch to fix those errors

Thanks @mcastorina for letting me. I have fixed the breaking changes. Seems like `Github Action` is facing downtime. I will wait for the check queues to get completed.

abmussani committed:
Merge branch 'main' into impl-data-model-gitlab
